tiledb/common/thread_pool/thread_pool.h

```diff
@@ -47,7 +47,241 @@ namespace tiledb::common {
 
 class ThreadPool {
  public:
-  using Task = std::future<Status>;
+  /**
+   * @brief Abstract base class for tasks that can run in this threadpool.
+   */
+  class ThreadPoolTask {
+   public:
+    ThreadPoolTask() = default;
+    ThreadPoolTask(ThreadPool* tp)
+        : tp_(tp){};
+    virtual ~ThreadPoolTask(){};
+
+   protected:
+    friend class ThreadPool;
+
+    ThreadPool* tp_{nullptr};
```

Comment:
I think it's ok since it's default constructed and therefore applies to all constructors. As C.41 mentions:
`- (Simple) Every constructor should initialize every data member (either explicitly, via a delegating ctor call or via default construction).`